### PR TITLE
DB level check for text field choices

### LINKED ISSUE #1046: Char/text field Choices are not enforced/validated at DB level
```
$ http https://PULP3-SOURCE-FEDORA36.puffy.example.com/pulp/api/v3/content/container/manifests/ --auth admin:password
HTTP/1.1 200 OK
Access-Control-Expose-Headers: Correlation-ID
Allow: GET, HEAD, OPTIONS
Connection: keep-alive
Content-Length: 722
Content-Type: application/json
Correlation-ID: 981cc643c6084edf9cf8950120187798
Date: Thu, 08 Sep 2022 14:47:26 GMT
Referrer-Policy: same-origin
Server: nginx
Vary: Accept, Cookie
X-Content-Type-Options: nosniff
X-Frame-Options: DENY

{
    "count": 1,
    "next": null,
    "previous": null,
    "results": [
        {
            "artifact": "/pulp/api/v3/artifacts/cf459685-dc38-435c-ab1c-07865b28b92d/",
            "blobs": [
                "/pulp/api/v3/content/container/blobs/ed537584-ab98-47d6-9b06-1fa366389ed8/",
                "/pulp/api/v3/content/container/blobs/6629a8b9-4ac8-4698-b10c-dc94e7d25f22/",
                "/pulp/api/v3/content/container/blobs/4c375c8b-5a49-4ad5-a226-b371ac2b6192/"
            ],
            "config_blob": null,
            "digest": "sha256:3bf3f08a5c4ca0b04ecdc0a95fa11530f86f36bd9d372a568254680e6332a2c0",
            "listed_manifests": [],
            "media_type": "application/vnd.docker.distribution.manifest.v1+prettyjws",
            "pulp_created": "2022-09-08T14:47:17.896191Z",
            "pulp_href": "/pulp/api/v3/content/container/manifests/fd2cae74-7fe1-4c8a-ab3c-00a06d6fdc55/",
            "schema_version": 1
        }
    ]
}


(pulp) [vagrant@pulp3-source-fedora36 ~]$ django-admin shell_plus
# Shell Plus Model Imports
from pulpcore.app.models.access_policy import AccessPolicy, Group
from django.contrib.auth.models import Permission, User
from django.contrib.contenttypes.models import ContentType
from django.contrib.sessions.models import Session
from pulp_ansible.app.models import AnsibleCollectionDeprecated, AnsibleDistribution, AnsibleRepository, Collection, CollectionImport, CollectionRemote, CollectionVersion, CollectionVersionSignature, GitRemote, RoleRemote
from pulpcore.app.models.role import GroupRole, Role, UserRole
from pulp_container.app.models import Blob, BlobManifest, ContainerDistribution, ContainerNamespace, ContainerPushRepository, ContainerRemote, ContainerRepository, Manifest, ManifestListManifest, ManifestSignature, ManifestSigningService, Tag
from pulpcore.app.models.upload import Upload, UploadChunk
from pulp_file.app.models import FileAlternateContentSource, FileContent, FileDistribution, FilePublication, FileRemote, FileRepository
from pulp_rpm.app.models.acs import RpmAlternateContentSource
from pulp_rpm.app.models.advisory import UpdateCollection, UpdateCollectionPackage, UpdateRecord, UpdateReference
from pulp_rpm.app.models.comps import PackageCategory, PackageEnvironment, PackageGroup, PackageLangpacks
from pulp_rpm.app.models.custom_metadata import RepoMetadataFile
from pulp_rpm.app.models.distribution import Addon, Checksum, DistributionTree, Image, Variant
from pulp_rpm.app.models.modulemd import Modulemd, ModulemdDefaults, ModulemdObsolete
from pulp_rpm.app.models.package import Package
from pulp_rpm.app.models.repository import RpmDistribution, RpmPublication, RpmRemote, RpmRepository, UlnRemote
from pulpcore.app.models.acs import AlternateContentSource, AlternateContentSourcePath
from pulpcore.app.models.base import Label
from pulpcore.app.models.content import Artifact, AsciiArmoredDetachedSigningService, Content, ContentArtifact, PulpTemporaryFile, RemoteArtifact, SigningService
from pulpcore.app.models.exporter import Export, ExportedResource, Exporter, FilesystemExport, FilesystemExporter, PulpExport, PulpExporter
from pulpcore.app.models.importer import Import, Importer, PulpImport, PulpImporter, PulpImporterRepository
from pulpcore.app.models.progress import GroupProgressReport, ProgressReport
from pulpcore.app.models.publication import BaseDistribution, ContentGuard, ContentRedirectContentGuard, Distribution, Publication, PublishedArtifact, PublishedMetadata, RBACContentGuard
from pulpcore.app.models.repository import Remote, Repository, RepositoryContent, RepositoryVersion, RepositoryVersionContentDetails
from pulpcore.app.models.status import ContentAppStatus
from pulpcore.app.models.task import CreatedResource, Task, TaskGroup, TaskSchedule, Worker
from pulpcore.app.models.telemetry import SystemID
# Shell Plus Django Imports
from django.core.cache import cache
from django.conf import settings
from django.contrib.auth import get_user_model
from django.db import transaction
from django.db.models import Avg, Case, Count, F, Max, Min, Prefetch, Q, Sum, When
from django.utils import timezone
from django.urls import reverse
from django.db.models import Exists, OuterRef, Subquery
Python 3.10.5 (main, Jun  9 2022, 00:00:00) [GCC 12.1.1 20220507 (Red Hat 12.1.1-1)]
Type 'copyright', 'credits' or 'license' for more information
IPython 8.4.0 -- An enhanced Interactive Python. Type '?' for help.

In [1]: Manifest.objects.first()
Out[1]: <Manifest: pk=fd2cae74-7fe1-4c8a-ab3c-00a06d6fdc55>


In [3]: Manifest.objects.first().media_type
Out[3]: 'application/vnd.docker.distribution.manifest.v1+prettyjws'

In [4]: m=Manifest.objects.first()

In [5]: m
Out[5]: <Manifest: pk=fd2cae74-7fe1-4c8a-ab3c-00a06d6fdc55>

In [6]: m.media_type
Out[6]: 'application/vnd.docker.distribution.manifest.v1+prettyjws'


In [10]: m.media_type='application/vnd.docker.distribution.manifest.v5+prettyjws'

In [11]: m.save()

In [12]: m=Manifest.objects.first()

In [13]: m.media_type
Out[13]: 'application/vnd.docker.distribution.manifest.v5+prettyjws'

In [14]: 

In [14]: 

In [14]: m.media_type='blablabla'

In [15]: m.save()

In [16]: m=Manifest.objects.first()

In [17]: m.media_type
Out[17]: 'blablabla'

In [18]:                                                                                                                                                                                                                                      
Do you really want to exit ([y]/n)? ^[[A^[[B^C

In [18]:                                                                                                                                                                                                                                      
Do you really want to exit ([y]/n)? y
(pulp) [vagrant@pulp3-source-fedora36 ~]$ http https://PULP3-SOURCE-FEDORA36.puffy.example.com/pulp/api/v3/content/container/manifests/ --auth admin:password
HTTP/1.1 200 OK
Access-Control-Expose-Headers: Correlation-ID
Allow: GET, HEAD, OPTIONS
Connection: keep-alive
Content-Length: 674
Content-Type: application/json
Correlation-ID: 623478c99c6a4b28908be1e57d69205a
Date: Thu, 08 Sep 2022 14:57:08 GMT
Referrer-Policy: same-origin
Server: nginx
Vary: Accept, Cookie
X-Content-Type-Options: nosniff
X-Frame-Options: DENY

{
    "count": 1,
    "next": null,
    "previous": null,
    "results": [
        {
            "artifact": "/pulp/api/v3/artifacts/cf459685-dc38-435c-ab1c-07865b28b92d/",
            "blobs": [
                "/pulp/api/v3/content/container/blobs/ed537584-ab98-47d6-9b06-1fa366389ed8/",
                "/pulp/api/v3/content/container/blobs/6629a8b9-4ac8-4698-b10c-dc94e7d25f22/",
                "/pulp/api/v3/content/container/blobs/4c375c8b-5a49-4ad5-a226-b371ac2b6192/"
            ],
            "config_blob": null,
            "digest": "sha256:3bf3f08a5c4ca0b04ecdc0a95fa11530f86f36bd9d372a568254680e6332a2c0",
            "listed_manifests": [],
            "media_type": "blablabla",
            "pulp_created": "2022-09-08T14:47:17.896191Z",
            "pulp_href": "/pulp/api/v3/content/container/manifests/fd2cae74-7fe1-4c8a-ab3c-00a06d6fdc55/",
            "schema_version": 1
        }
    ]
}


```
https://github.com/pulp/pulp_container/blob/main/pulp_container/app/models.py#L84

### PR DESCRIPTION
**Pull request by MichalPysik**
Some models in pulp_container have a text field, whose value must be one of the predefined choices. This check was only done on the application level, but it is now done on the database level too.

closes #1046

### CODE REVIEW

**Comment by ipanova** [comment]
did you test the upgrade scenario? how this is going to be handled for already existing and saved records with random data that is not part of choces?

File: CHANGES/1046.bugfix [hunk 1]
```diff
@@ -0,0 +1,2 @@
+Added database-level validation for the 'media_type' field of Manifest and the 'type' field of 
+Manifest and the 'type' field of ManifestSignature.
```
Comment (lubosmj) [inline CHANGES/1046.bugfix:2]:
```suggestion
Enforced database-level validation for the 'media_type' field of Manifest and the 'type' field of ManifestSignature.
```

**Comment by lubosmj** [comment]
> Also, I haven't put that much thought into the constraint names, and a custom error code and error message can also be specified if needed https://docs.djangoproject.com/en/5.0/ref/models/constraints/#checkconstraint , so we can discuss that as well.

I am fine with the names of the constraints. I think there is no need to customize the error message. How does it look like now when a user saves a manifest with an invalid media type?

And, yes, definitely, try the scenario Ina sketched.

**Comment by lubosmj** [comment]
> did you test the upgrade scenario? how this is going to be handled for already existing and saved records with random data that is not part of choices?

Is there a way to ship a **migration script that detects invalid media types** and lists them back to a user? Then, I guess the user will need to manually update/remove affected manifests. Will she remove the associated blobs too? I have no clue how the potential failure of the migration would be handled.

**Comment by MichalPysik** [comment]
Since the migration would require the user to remove/change all database entries for manifests with "incorrect" media_type, we decided to close this PR (and issue), since even the app-level validation for this field is planned to go away. This refers to point 4 here: https://github.com/pulp/pulp_container/issues/1494 .

**Pull request closed without merging**